File: Part1_강의자료/Ch07_이분탐색/문제별코드/2295_세수의합/2295_set.java
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

class Main
{
    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int[] arr = new int[N];
        for (int i = 0; i < N; i++)
            arr[i] = sc.nextInt();

        Set<Integer> sums = new HashSet<>();
        for (int i = 0; i < N; i++)
            for (int j = i; j < N; j++)
                sums.add(arr[i] + arr[j]);

        int ans = -1;
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++) {
                int target = arr[i] - arr[j];
                if (sums.contains(target))
                    ans = Math.max(ans, arr[i]);
            }
        System.out.println(ans);
    }
}

File: Part1_강의자료/Ch07_이분탐색/문제별코드/2295_세수의합/2295_set.py
N = int(input())
arr = sorted([int(input()) for _ in range(N)])

sums = set()
for i in range(N):
    for j in range(i, N):
        sums.add(arr[i] + arr[j])

flag = False
for i in range(N - 1, 0, -1):
    for j in range(i - 1, -1, -1):
        if arr[i] - arr[j] in sums:
            print(arr[i])
            flag = True
            break
    if flag is True:
        break
